File: package/ios/React/FpsSampleCollector.swift
//
//  FpsSampleCollector.swift
//  VisionCamera
//
//  Created by Marc Rousavy on 22.04.24.
//

import Foundation

final class FpsSampleCollector {
  private var timestamps: [UInt64] = []
  private var timer: Timer?
  weak var delegate: FpsSampleCollectorDelegate?

  func start() {
    timer = Timer.scheduledTimer(withTimeInterval: 1, repeats: true, block: { [weak self] timer in
      guard let self else {
        timer.invalidate()
        return
      }
      guard let delegate = self.delegate else { return }
      delegate.onAverageFpsChanged(averageFps: self.averageFps)
    })
  }

  func stop() {
    timer?.invalidate()
    timer = nil
  }

  /**
   Add a new timestamp to the FPS samples.
   */
  func onTick() {
    let now = DispatchTime.now().uptimeNanoseconds
    timestamps.append(now)

    // filter out any timestamps that are older than 1 second
    timestamps = timestamps.filter { timestamp in
      let differenceMs = (now - timestamp) / 1_000_000
      return differenceMs < 1000
    }
  }

  /**
   Gets the current average FPS.
   */
  var averageFps: Double {
    guard let firstTimestamp = timestamps.first,
          let lastTimestamp = timestamps.last else {
      return 0.0
    }

    let totalDurationMs = (lastTimestamp - firstTimestamp) / 1_000_000
    let averageFrameDurationMs = Double(totalDurationMs) / Double(timestamps.count - 1)

    return 1000.0 / averageFrameDurationMs
  }
}

File: android/src/main/java/com/mrousavy/camera/react/FpsSampleCollector.kt
package com.mrousavy.camera.react

import java.util.Timer
import kotlin.concurrent.schedule

class FpsSampleCollector(val callback: Callback) {
  private var timestamps = mutableListOf<Long>()
  private var timer: Timer? = null

  fun start() {
    timer = Timer("VisionCamera FPS Sample Collector")
    timer?.schedule(1000, 1000) {
      callback.onAverageFpsChanged(averageFps)
    }
  }

  fun stop() {
    timer?.cancel()
    timer = null
  }

  fun onTick() {
    val now = System.currentTimeMillis()
    timestamps.add(now)

    // filter out any timestamps that are older than 1 second
    timestamps = timestamps.filter { timestamp ->
      val differenceMs = now - timestamp
      return@filter differenceMs < 1000
    }.toMutableList()
  }

  private val averageFps: Double
    get() {
      val first = timestamps.firstOrNull()
      val last = timestamps.lastOrNull()
      if (first == null || last == null) return 0.0

      val totalDurationMs = last - first
      val averageFrameDurationMs = totalDurationMs.toDouble() / (timestamps.count() - 1).toDouble()
      return 1000.0 / averageFrameDurationMs
    }

  interface Callback {
    fun onAverageFpsChanged(averageFps: Double)
  }
}
